test/parachain/src/cli.rs

```diff
@@ -95,22 +98,18 @@ where
 				config.network.listen_addresses = Vec::new();
 
 				// TODO
-				let mut polkadot_config =
-					polkadot_collator::Configuration::default_with_spec_and_base_path(
-						polkadot_service::ChainSpec::from_json_bytes(
-							&include_bytes!("../res/polkadot_chainspec.json")[..],
-						)?,
-						config.in_chain_config_dir("polkadot"),
-					);
+				let mut polkadot_config = parse_and_prepare::<NoCustom, NoCustom, _>(
+					&version,
+					"cumulus-test-parachain-collator",
+					args_relaychain,
+				).into_configuration::<CustomConfigurationPolkadot, _, _, _>(
+					load_spec_polkadot,
+					config.in_chain_config_dir("polkadot"),
+				)
+					.map_err(|e| e.to_string())?
+					.expect("not a run command?");
```

Comment:
in Substrate and Polkadot codebases, we have a stricter policy on panickers than that - please provide a logical proof (premise -> conclusion) that this does not fail.

Reply:
I don't think I can proof anything here. The user can send other commands than RunCmd. But I think we expect only RunCmd here, right? @bkchr

Reply:
If you can't prove anything, then you can get rid of the `expect` and exit gracefully.

Reply:
Yeah, just do `ok_or_else()?`

Reply:
I just checked because I changed the code and I forgot exactly what was the issue. But I left that "?" expressly for someone to check that at the review.

This is an unwrap on an Option. This option is always Some unless the command was a custom command. In the current context, it is impossible that it is a custom command because the `parse_and_prepare` is ran with NoCustom. So this won't happen.

Nevertheless... the subcommand can be anything: it can be a RunCmd (what we expect) but also an import-blocks, export-blocks and everything else that is supposed. This makes no sense and I'm afraid at the comment there isn't much I can't do to prevent this.

Reply:
> This is an unwrap on an Option. This option is always Some unless the command was a custom command. In the current context, it is impossible that it is a custom command because the parse_and_prepare is ran with NoCustom. So this won't happen.

The point is to put this logic _in the code_ so that if it triggers later we can see exactly what assumption was broken. I am not saying to never use `expect`, only to use it with our guidelines.

Here is an example :

`expect("can only fail when this is not a RunCmd. Running parse_and_prepare with NoCustom can only return a RunCmd; therefore this is a RunCmd; qed")`

Reply:
@rphmeier I adapted the message consequently but I think you misunderstood what I said.

Whatever I do, the user CAN use import-blocks, export-blocks and the other subcommands in the relay chain arguments. There is nothing I can do to prevent that unless I do quite big modifications on substrate's cli.

I believe the current behavior is acceptable for now but I plan to refactor substrate's cli and I will fix this in a near future.

(The current behavior is: whatever subcommand you use in the relay chain arguments, it will run the collator)

Reply:
> Whatever I do, the user CAN use import-blocks, export-blocks and the other subcommands in the relay chain arguments. There is nothing I can do to prevent that unless I do quite big modifications on substrate's cli.

I followed you this far in the previous comment, and I agree that it would be a large refactoring. Can you clarify: the `expect` call you added cannot be triggered by user input?

Reply:
No it's impossible. `into_configuration` returns None only in case of a custom command but parse_and_prepare (the second one, not the first one) is ran without custom command (NoCustom)

Reply:
I have already started the refactoring and it will makes some things (like this) much easier: https://github.com/paritytech/substrate/pull/4692